Some enhancements to slack-rtm-api

robzr opened the pull request:
Hi Rémi, I like your slack-rtm-api module, and am planning on using it in a Slack Bot framework I'm working on.  I made some changes to it, if you are interested in merging, check them out.
- Added support for Slack url_reconnects and automatic connection re-establishment
- Added a WebSocket::Driver on :close handler (for automatic connection re-establishment)
- Changed the socket read to not block by using IO::select
- Changed the initialize arguments to use a Hash with symbol keys (I think this is the only not-backwards compatible change)
- I forgot to put this in the git commit, but I added support for WebSocket ping/pong

I'm new to github and pull requests so bear with me if I'm breeching any etiquette or not doing something right.  Thanks,

Rob

robzr committed:
Added require 'net/http' to avoid error on OS X

robzr committed:
Changed timing of send event queue, added Slack reconnect_url handling

Changed timing of the send event queue so that it is checked even when
there is nothing to read on the socket.

Added reconnect_url handling to the WebSocket::Driver on :message
callback to update the websocket @url

robzr committed:
Added reconnect handler via @driver.on :close in init

robzr committed:
Changed initialize args to hash, added start by default

Changed arguments on initialize to use hash with symbols
Now starts on instantiation unless start: false is passed

robzr committed:
Took out auto-start as it does not work :)

robzr committed:
Added IO::select around socket read to avoid blocking

rdlh commented:
Hey @robzr !

Thank you for this PR, I'll review it tomorrow (it's 11:42 PM here).

Cheers,
Rémi

robzr committed:
Updated connection tracking, added #close method

With values showing the state of transition
:closed => connection is closed
:connecting => websocket is connecting but not opened yet
:ini =>
:open =>

robzr committed:
Updated connection tracking, #close method

With values showing the state of transition
:closed => connection is closed
:connecting => websocket is connecting but not opened yet
:inializing => connection opened to slack but has not finished handshake
:open => handshake is done, connection is good to go

close method will kill the thread, gracefully shut down websocket, set
status to :closed

robzr committed:
Made IO select timeout a passable variable

robzr committed:
Added << method alias to send event, misc changes

robzr committed:
Fixed type-o in attr_accesor

robzr committed:
Added wait on open functionality with timeout

rdlh commented:
@robzr 

PR is merged, thank you again!
Can you please **update README** in another PR?
